etlman/projects/tests/test_models.py

```diff
@@ -16,6 +18,42 @@ def test_django_model_str_methods(self):
         builded_object = ProjectFactory.build()
         assert builded_object.name == str(builded_object)
         builded_object = PipelineFactory.build()
-        assert builded_object.name == str(builded_object)
+        assert f"{builded_object.name}, pk: {builded_object.id}" == str(builded_object)
         builded_object = StepFactory.build()
         assert builded_object.name == str(builded_object)
+
+
+@pytest.mark.django_db
+class TestPipelineModel:
+    def test_run_pipeline_single_step(self):
+        """
+        Pipeline.run_pipeline() method creates the correct PipelineRun object in the database.
+        """
+        pipeline = PipelineFactory()
+        step = StepFactory(pipeline=pipeline)
+        backend = TestBackend()
+        pipeline.run_pipeline(backend=backend)
+        PipelineRun.objects.count() == 1
+        last_pipeline_run = PipelineRun.objects.all().first()
+        assert pipeline == last_pipeline_run.pipeline
+        assert last_pipeline_run.output["pipeline_id"] == pipeline.pk
+        assert last_pipeline_run.output["steps"][0]["step_id"] == step.pk
+        assert last_pipeline_run.output["steps"][0]["stdout"].startswith(
+            TestBackend.STDOUT_MARKER
+        )
+        assert last_pipeline_run.output["steps"][0]["stderr"].startswith(
+            TestBackend.STDERR_MARKER
+        )
+
+    def test_run_pipeline_multiple_steps(self):
+        """
+        Pipeline.run_pipeline() method creates the correct PipelineRun object in the database
+        for a Pipeline with multiple Steps.
+        """
+        pipeline = PipelineFactory()
+        StepFactory(pipeline=pipeline)
+        StepFactory(pipeline=pipeline)
+        StepFactory(pipeline=pipeline)
+        backend = TestBackend()
+        pipeline.run_pipeline(backend=backend)
+        PipelineRun.objects.count() == 3
```

Comment:
This line is missing an `assert`, and fails when the assert is added. There should only be one `PipelineRun` in this case. I added some sample additional things we could look for in the 'single' version of this test. Could you expand this test to look for the following in the pipeline run `output`?

- The correct `pipeline_id`
- The correct number of steps
- Bonus: The correct step IDs (for this one, use a python `set()` to compare the IDs since they might not come back in the same order from the database as they appear in the `steps` list in `output`)

Since we look for the stderr/stdout markers in the 'single' version of this test, we can skip that here. We mainly want to look for any issues that might exist with going from one -> multiple steps.

Reply:
Great suggestions. I was not aware of `.output`.
They are all implemented.